OpenRA.Mods.Cnc/Widgets/Logic/CampaignLogic/CampaignMissionBrowserLogic.cs

```diff
@@ -0,0 +1,268 @@
+			campaignPreviewGraficButton.OnClick = campaignWorld.CallbackShowCampaignBrowserOnClick;
+
+			// Campaign preview button
+			campaignPreviewContinueButton = widget.Get<ButtonWidget>("CAMPAIGN_PREVIEW_CONTINUE_BUTTON");
+			campaignPreviewContinueButton.OnClick = campaignWorld.CallbackShowCampaignBrowserOnClick;
+
+			// Campaign preview back button
+			campaignBrowserBackButton = widget.Get<ButtonWidget>("CAMPAIGN_PREVIEW_BACK_BUTTON");
+			campaignBrowserBackButton.OnClick = () =>
+			{
+				if (this.congratsFlag)
+					campaignWorld.ShowCongratulations();
+				else
+				{
+					Game.Disconnect();
+					Ui.CloseWindow();
+					onExit();
+				}
+			};
+		}
+
+		public bool GetCampaignPreviewRequired()
+		{
+			return this.campaignPreviewRequired;
+		}
+
+		void ConfigureNextAndPrevButton()
+		{
+			if (nextMaps.Count > 1)
+			{
+				nextButton.OnClick = () =>
+				{
+					mapIndex = (mapIndex + 1) % nextMaps.Count;
+					nextMap = nextMaps[mapIndex];
+					campaignWorld.SetMapContent();
+				};
+
+				prevButton.OnClick = () =>
+				{
+					mapIndex = (mapIndex + nextMaps.Count - 1) % nextMaps.Count;
+					nextMap = nextMaps[mapIndex];
+					campaignWorld.SetMapContent();
+				};
+			}
+			else
+			{
+				nextButton.Visible = false;
+				prevButton.Visible = false;
+			}
+		}
+
+		public void ProgressCampaign()
+		{
+			this.LoadFactionMaps();
+			this.LoadCurrentMissions();
+			this.CheckCampaignCompleted();
+			this.ConfigureNextAndPrevButton();
+
+			this.CheckCampaignProgressForPreview();
+
+			if ((CampaignWorldLogic.Campaign.Equals("GDI Campaign") && lastMissionSuccessfullyPlayedGDI) ||
+				(CampaignWorldLogic.Campaign.Equals("Nod Campaign") && lastMissionSuccessfullyPlayedNod))
+				campaignWorld.ShowCongratulations();
+			else if (campaignPreviewRequired)
+				campaignWorld.ShowCampaignPreview();
+			else
+				campaignWorld.CallbackShowCampaignBrowserOnClick();
+		}
+
+		public void CheckCampaignProgressForPreview()
+		{
+			// Check campaign preview
+			if (nextMaps.Count() > 1)	// More missions available - campaign preview required
+				campaignPreviewRequired = true;
+			else
+				campaignPreviewRequired = false;
+		}
+
+		public Map GetNextMap()
+		{
+			return this.nextMap;
+		}
+
+		public MapPreview GetSelectedMapPreview()
+		{
+			return this.selectedMapPreview;
+		}
+
+		public void SetSelectedMapPreview(MapPreview selectedMapPreview)
+		{
+			this.selectedMapPreview = selectedMapPreview;
+		}
+
+		public void SetPreviewContent()
+		{
+			if (this.GetNextMap().Container.Exists("preview.png"))
+				using (var dataStream = this.GetNextMap().Container.GetContent("preview.png"))
+				{
+					this.GetNextMap().CustomPreview = new System.Drawing.Bitmap(dataStream);
+					var preview = new MapPreview(this.GetNextMap().Uid, Game.ModData.MapCache);
+					preview.SetMinimap(new SheetBuilder(SheetType.BGRA).Add(this.GetNextMap().CustomPreview));
```

Comment:
Use the MapCache to generate your preview image.  Doing this manually is very inefficient.